src/ui/features/xp-drop/XpDropClaimFlow/XpDropClaim.tsx

```diff
@@ -0,0 +1,510 @@
+  const navigate = useNavigate();
+
+  const { data: currentWallet } = useQuery({
+    queryKey: ['wallet/uiGetCurrentWallet'],
+    queryFn: () => walletPort.request('uiGetCurrentWallet'),
+  });
+  const [selectedWallet, setSelectedWallet] = useState(currentWallet);
+
+  const { data: walletGroups, ...walletGroupsQuery } = useWalletGroups();
+  const { signerWalletGroups, signerWalletAddresses } = useMemo(() => {
+    const signerWalletGroups =
+      walletGroups?.filter((group) =>
+        isSignerContainer(group.walletContainer)
+      ) ?? [];
+    const signerWalletAddresses = signerWalletGroups
+      .flatMap((group) => group.walletContainer.wallets)
+      .map((wallet) => wallet.address);
+    return { signerWalletGroups, signerWalletAddresses };
+  }, [walletGroups]);
+
+  const { data: walletsMeta, ...walletsMetaQuery } = useQuery({
+    queryKey: ['ZerionAPI.getWalletsMeta', signerWalletAddresses],
+    queryFn: () => getWalletsMetaByChunks(signerWalletAddresses),
+  });
+
+  const { eligibleWalletGroups, eligibleAddresses } = useMemo(() => {
+    const eligibleAddresses =
+      walletsMeta
+        ?.filter((meta) => Boolean(meta.membership.retro))
+        .map((meta) => normalizeAddress(meta.address)) ?? [];
+
+    const eligibleWalletGroups = signerWalletGroups
+      ?.map((group) => ({
+        id: group.id,
+        walletContainer: {
+          wallets: group.walletContainer.wallets.filter((wallet) =>
+            eligibleAddresses.includes(normalizeAddress(wallet.address))
+          ),
+        },
+      }))
+      .filter((group) => group.walletContainer.wallets.length > 0);
+
+    return { eligibleWalletGroups, eligibleAddresses };
+  }, [signerWalletGroups, walletsMeta]);
+
+  const signMsgBtnRef = useRef<SignMsgBtnHandle | null>(null);
+  const walletSelectDialogRef = useRef<HTMLDialogElementInterface | null>(null);
+  const xpBreakdownDialogRef = useRef<HTMLDialogElementInterface | null>(null);
+
+  const { mutate: personalSign, ...personalSignMutation } = useMutation({
+    mutationFn: async () => {
+      invariant(selectedWallet, 'selectedWallet is required');
+      invariant(signMsgBtnRef.current, 'SignMessageButton not found');
+
+      const message = 'I want to claim my Zerion Retro XP';
+      const signature = await signMsgBtnRef.current.personalSign({
+        params: [message],
+        initiator: INTERNAL_ORIGIN,
+        clientScope: null,
+      });
+      await ZerionAPI.claimRetro({
+        address: selectedWallet.address,
+        signature,
+      });
+      // Typically, after we call `/wallet/claim-retro/v1` it takes asecond or two
+      // before `membership.retro` becomes `null` (in response to `/wallet/get-meta/v1`).
+      // To account for this possible delay, we're adding a short wait here.
+      // Otherwise, we might still see the wallet for which we just claimed the XP drop in
+      // the wallet selector when we click on the "Next Wallet" button.
+      await wait(2000);
```

Comment:
Ideally, we should save this wallet locally (at least, for the session) and don't show it again after successful claim.
I don't mind waiting here, but it is better not to wait for backend to update.

Reply:
I like your suggestion, I'll try smth like this:
```
  const [claimedWallet, setClaimedWallet] = useState(null);
  ...
  const { eligibleWalletGroups, eligibleAddresses } = useMemo(() => {
    ... 
    return eligibleAddresses.includes(normalizedAddress) && normalizedAddress !== claimedWallet?.address;
  });
  ...
  setClaimedWalet(wallet);
```

Reply:
ok, I completely changed the way I manage state in this component ✅